### DISCUSSION TITLE
Linear algebra in Python 3

### DISCUSSION THREAD

**Discussion by nctrd** [Homebrew/discussions#1701] [Everyday usage]
Dear Developers, 

I have a Homebrew Python 3 installation on my Mac OS, from this formula:
https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/python@3.9.rb

On top of that, I have a pip3-installed Numpy and Scipy. 
They link to libraries in /usr/local/lib/python3.9/site-packages/numpy/.dylibs/, one of which is libopenblas.0.dylib.
My attempts to get linear algebra version from inside Numpy indicate that it does not link to Accelerate (Mac OS linalg) and does not support it. 
Hence, my Numpy apparently comes with its own full-sized openBLAS (~60MB binary). 
At least, I think it is an openBLAS binary. 

The question is: how do I get its version?

With otool -L I got this:
/usr/local/lib/python3.9/site-packages/numpy/core/../.dylibs/libopenblas.0.dylib:
    /DLC/numpy/libopenblas.0.dylib (**compatibility version 0.0.0, current version 0.0.0)**  
    @loader_path/libgfortran.3.dylib (compatibility version 4.0.0, current version 4.0.0)
    /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
    @loader_path/libquadmath.0.dylib (compatibility version 1.0.0, current version 1.0.0)
    @loader_path/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.0.0)

Also, please forgive me if this is the wrong place to ask. 

Thank you in advance. 

Best regards. 
nctrd

**Comment by SMillerDev** [comment]
> The question is: how do I get its version?

A quick search for `numpy get config` reveals: https://betterscientificsoftware.github.io/python-for-hpc/interrogating-numpy/
Which suggests: 

```python
import numpy
numpy.__config__.show()
```

**Reply by nctrd** [reply]
Dear Sean, 

That's what I did. It shows /usr/local/lib, basically.
I then did:
```
import numpy.distutils.system_info as sysinfo
sysinfo.get_info('lapack')
```

which gave me:
`RuntimeError: Found /usr/lib/liblapack.dylib, but that file is a symbolic link to the MacOS Accelerate framework, which is not supported by NumPy. You must configure the build to use a different optimized library, or disable the use of optimized BLAS and LAPACK by setting the environment variables NPY_BLAS_ORDER="" and NPY_LAPACK_ORDER="" before building NumPy.`

So, there's that.

**Reply by cho-m** [reply]
Sounds like an issue with linking libraries during installation.

Maybe when installing `numpy`, you can try to set linkage to use supported libraries (like Homebrew's `openblas` formula).

For example, something similar to Homebrew's [`numpy` formula](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/numpy.rb#L24-L33) where the `ATLAS`, `BLAS`, and `LAPACK` variables are set. Maybe also the config file too.

Or perhaps directly setting `OPENBLAS` environment variable, e.g.
```
brew install openblas
OPENBLAS="$(brew --prefix openblas)" pip install numpy
```

**Comment by carlocab** [comment]
A `pip`-installed `numpy` doesn't seem to make it easy for you to query the version of the bundled `libopenblas`.

You _might_ be able to write a small C program that links to the bundled `libopenblas` and then queries it for the version, but I don't know if OpenBLAS exposes an API that allows you to do this.
